### PR TITLE
Add market supply caps

### PR DESCRIPTION
**Pull request by StanleyDing**

### CODE REVIEW

File: contracts/Comptroller.sol [hunk 1]
```diff
@@ -245,12 +251,28 @@ contract Comptroller is ComptrollerV4Storage, ComptrollerInterface, ComptrollerE
...
 
+        uint supplyCap = supplyCaps[cToken];
+        // Supply cap of 0 corresponds to unlimited supplying
+        if (supplyCap != 0) {
+            uint totalCash = CToken(cToken).getCash();
+            uint totalBorrows = CToken(cToken).totalBorrows();
+            uint totalReserves = CToken(cToken).totalReserves();
+            // totalSupplies = totalCash + totalBorrows - totalReserves
+            (MathError mathErr, uint totalSupplies) = addThenSubUInt(totalCash, totalBorrows, totalReserves);
+            require(mathErr == MathError.NO_ERROR, "totalSupplies failed");
```
Comment (bun919tw) [inline contracts/Comptroller.sol:267]: minor: why is this camelcase?

**Review by bun919tw — APPROVED** [review]
LGTM, please let @0xyolo double-check the scenario tests.

File: contracts/Comptroller.sol (hunk 1 above)
Comment (StanleyDing) [reply contracts/Comptroller.sol:267]: 因為我是指那個變數

**Comment by bun919tw** [comment]
Comptroller has become too large to deploy. Add two commits to toggle the problem.
Credits to [compound's PR](https://github.com/compound-finance/compound-protocol/pull/71).
Also here's a good reference by [compound](https://www.comp.xyz/t/a-leaner-meaner-comptroller/461).

- Add `mul_ScalarTruncate` and `mul_ScalarTruncateAddUint` to `Exponential.sol`.
- Add `_setCompSpeeds`.
- Remove `_setMaxAssets`.
- Remove `_setCompRate` and `refreshCompSpeeds`.
- Remove checks in `_setCloseFactor` and `_setLiquidationIncentive`.

**Review by jeremyHD — APPROVED** [review]

File: contracts/Comptroller.sol (hunk 1 above)
Comment (ihateaher0) [reply contracts/Comptroller.sol:244]: kk

**Pull request merged**